### PR TITLE
Enhance the tests of `NSGAIISampler._crowding_distance_sort`

### PR DESCRIPTION
**Pull request by knshnb**
## Motivation

The current test of `NSGAIISampler._crowding_distance_sort` is very limited. In this PR, I refactored the existing code and enhanced the test.

It seems that currently, inputs of `float("inf")` or `float("nan")` are not considered. I will create a separate issue for that.

## Description of the changes

- split `_calc_crowding_distance` from `_crowding_distance_sort`
- added test for each function
- added edge-case inputs

### CODE REVIEW

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. Basically, LGTM. I have a comment on the test of `_calc_distance_sort`. PTAL.

File: tests/samplers_tests/test_nsgaii.py [hunk 1]
```diff
@@ -419,24 +419,49 @@ def test_fast_non_dominated_sort_missing_constraint_values() -> None:
...
+        (
+            [[1], [2], [float("inf")]],
+            [float("inf"), float("nan"), float("inf")],
+        ),  # TODO(knshnb): Decide expected behavior for this case.
+        (
+            [[0], [1], [float("nan")]],
+            [float("inf"), float("nan"), float("inf")],
+        ),  # TODO(knshnb): Decide expected behavior for this case.
+    ],
+)
```
[the remark this answers is not in the document]
Comment (knshnb) [reply tests/samplers_tests/test_nsgaii.py:442]: Thanks for pointing it out! I agree with the idea that we should only consider the cases that are actually used.
Let me add a comment to the original function on the assumption and modify the test case to meet that.

[the remark this answers is not in the document]
Comment (knshnb) [reply tests/samplers_tests/test_nsgaii.py:442]: Having said that, I also think it might be still okay to keep the first test case `([[5], [6], [9], [0]], [6 / 9, 4 / 9, float("inf"), float("inf")])` as a simple demonstration of how `_calc_crowding_distance` works. (All values must be the same in one-dimensional cases if they have the same nondomination rank.)
What do you think?

[the remark this answers is not in the document]
Comment (contramundum53) [reply tests/samplers_tests/test_nsgaii.py:442]: I think it's okay to also test the cases that are actually never used if that makes testing more efficient.

[the remark this answers is not in the document]
Comment (HideakiImamura) [reply tests/samplers_tests/test_nsgaii.py:442]: I don't see how test cases that are not actually used make testing more efficient. Do you have any specific examples?

[the remark this answers is not in the document]
Comment (contramundum53) [reply tests/samplers_tests/test_nsgaii.py:442]: I mean that could make the test cases simpler and easier to read, or fewer tests could cover all branches. The first 1D test case is one example. That case never actually appears but is conceptually simple and easy to understand.

**Review by contramundum53 — APPROVED** [review]
LGTM!

File: tests/samplers_tests/test_nsgaii.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (knshnb) [reply tests/samplers_tests/test_nsgaii.py:442]: I agree with @contramundum53. `_calc_crowding_distance` is decomposed into a sum of distances in each dimension. Although one-dimensional input with different values itself never appears because of the nondomination rank assumption, it appears as a part of multi-dimensional input. So, I think it is efficient to test several one-dimensional cases along with multi-dimensional ones.

[the remark this answers is not in the document]
Comment (HideakiImamura) [reply tests/samplers_tests/test_nsgaii.py:442]: Thank you for your comments. I agree with you idea. Let me resolve this comment.

File: tests/samplers_tests/test_nsgaii.py [hunk 2]
```diff
@@ -419,24 +419,49 @@ def test_fast_non_dominated_sort_missing_constraint_values() -> None:
...
-    ]
+
+@pytest.mark.parametrize(
+    "values",
+    [
+        [[5], [6], [9], [0]],
+        [[5, 0], [6, 0], [9, 0], [0, 0]],
+        [[5, -1], [6, 0], [9, 1], [0, 2]],
+    ],
+)
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_nsgaii.py:457]: The `values` can be `+/-inf`. Should we add those cases here?

Comment (knshnb) [reply tests/samplers_tests/test_nsgaii.py]: Thanks for the comment! I added the test cases.

File: tests/samplers_tests/test_nsgaii.py [hunk 3]
```diff
@@ -419,24 +419,51 @@ def test_fast_non_dominated_sort_missing_constraint_values() -> None:
...
-    ]
+
+@pytest.mark.parametrize(
+    "values",
+    [
+        [[5], [6], [9], [0]],
+        [[5, 0], [6, 0], [9, 0], [0, 0]],
+        [[5, -1], [6, 0], [9, 1], [0, 2]],
+        [[1], [2], [float("inf")]],
+        [[0], [1], [float("nan")]],
```
Comment (HideakiImamura) [inline tests/samplers_tests/test_nsgaii.py:457]: The values of trials cannot be `nan`. I think we can skip it. In addition, we should add `-inf`.

Comment (knshnb) [reply tests/samplers_tests/test_nsgaii.py]: Thanks for pointing it out. Fixed!

**Review by HideakiImamura — APPROVED** [review]
LGTM!
